### PR TITLE
[MIG][8.0] replenishment_cost + product_cost_incl_bom

### PR DESCRIPTION
**Pull request by SodexisTeam**

### CODE REVIEW

**Comment by atchuthan** [comment]
@pedrobaeza: add "needs review" label for this PR.

**Comment by gurneyalex** [comment]
:+1:

File: README.md [hunk 1]
```diff
@@ -15,14 +15,14 @@ Available addons
...
 --- | --- | ---
+[product_cost_incl_bom](product_cost_incl_bom/) | 8.0.2.1.0 | Product Cost incl. BOM
 [product_replenishment_cost](product_replenishment_cost/) | 8.0.2.0.0 | Replenishment cost
 [product_standard_margin](product_standard_margin/) | 8.0.2.0.0 | Product Margin and Margin Rate
 
 Unported addons
 ---------------
 addon | version | summary
 --- | --- | ---
-[product_cost_incl_bom](product_cost_incl_bom/) | 1.0 (unported) | Product Cost incl. BOM
```
Comment (Garamotte) [inline README.md]: This file is updated automatically by an OCA script.

File: product_cost_incl_bom/models/product.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        return computed
+
+    @api.multi
+    def _get_replenishment_cost(self):
+        res = self._compute_purchase_price()
+        for product in self:
+            cost = 0
+            if product.id in res.keys():
+                cost = res[product.id]
+            product.replenishment_cost = cost
```
Comment (Garamotte) [inline product_cost_incl_bom/models/product.py]: Maybe `product.replenishment_cost = res.get(product.id, 0)` ?

File: product_replenishment_cost/models/__init__.py [hunk 3]
```diff
@@ -1,22 +1,6 @@
...
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2012 Alexandre Fayolle,Yannick Vaucher,Joël Grand-Guillaume,Camptocamp
+# © 2016 Sodexis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from . import product_product
+from . import product_product, product_template
```
Comment (Garamotte) [inline product_replenishment_cost/models/__init__.py]: Please split one import per line.

File: product_replenishment_cost/models/product_product.py [hunk 4]
```diff
@@ -28,14 +12,15 @@
...
-    @api.one
+    @api.multi
     @api.depends('product_tmpl_id.standard_price')
     def _get_replenishment_cost(self):
-        self.replenishment_cost = self.standard_price
+        for product in self:
+            product.replenishment_cost = product.standard_price
 
     replenishment_cost = fields.Float(
-        compute=_get_replenishment_cost, store=True,
```
Comment (Garamotte) [inline product_replenishment_cost/models/product_product.py:37]: Why did you remove the `store` attribute ?

File: product_replenishment_cost/models/product_template.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import openerp.addons.decimal_precision as dp
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    @api.multi
+    @api.depends('standard_price')
+    def _get_replenishment_cost(self):
+        replenishment_cost = 0.0
```
Comment (Garamotte) [inline product_replenishment_cost/models/product_template.py]: If a template has no variant, it will pick the value computed for the previous template in the loop.
You should initialize the variable in the for loop.

File: product_replenishment_cost/models/product_template.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        for template in self:
+            variants = template.mapped('product_variant_ids')
```
Comment (Garamotte) [inline product_replenishment_cost/models/product_template.py]: `variants = template.product_variant_ids` is enough here, mapped is useful when you work on multiple records, or through multiple fields.

**Review by Garamotte — CHANGES_REQUESTED** [review]

File: product_replenishment_cost/models/product_product.py (hunk 4 above)
Comment (atchuthan) [reply product_replenishment_cost/models/product_product.py:37]: I had 2 reasons to remove store=True
1. The replenishment cost field defined in product_replenishment_cost module with store=True, creates the field in database with the value of product's standard_price and the value is not recomputed at the time of product_cost_incl_bom module installation.
2. Replenishment cost needs to be changed for each change in bom lines of product, when there is a change in their replenishment cost value and it did not work in that case as the method was not triggered for store=True. (when we change the BOM directly, it did not trigger the compute method and it does not allow multi-level BOM)

**Comment by atchuthan** [comment]
@sylvain-garancher corrections done

**Review by gurneyalex — APPROVED** [review]

**Review by dvdhinesh — APPROVED** [review]
Code review :+1:

File: product_replenishment_cost/models/product_product.py (hunk 4 above)
Comment (atchuthan) [reply product_replenishment_cost/models/product_product.py:37]: @sylvain-garancher please check the points I have listed out.

Comment (Garamotte) [reply product_replenishment_cost/models/product_product.py:37]: Ok, I see, and I don't find a trigger for that. With the explanation, it's fine.

If the issue is with the other module, maybe you can leave `store=True` here, and explicitely change to `store=False` in the other module ? (with a comment explaining why)

**Comment by guewen** [comment]
Restarted Travis and I don't get why the builds are shown as failed whereas the builds exit with 0.

**Comment by legalsylvain** [comment]
closing.
replenishment_cost is available in V8 and in V10.

You can also review the PR for V12 here : https://github.com/OCA/margin-analysis/pull/50
